DS-3406: Sort communities and collections in-memory using a comparator

tomdesair opened the pull request:
Fix for https://jira.duraspace.org/browse/DS-3406 where the sorting of communities and collections happens in-memory and not by Hibernate.

Commit:
DS-3406: Sort communities and collections in-memory using a comparator

terrywbrady commented:
@kshepherd , could you review this one?  The Jira ticket references some of your work.

mwoodiupui commented on dspace-api/src/test/java/org/dspace/content/comparator/NameAscendingComparatorTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Test
+    public void testCompareSecondNull() throws Exception {
+        when(dso2.getName()).thenReturn("b");
```
Should that be "dso1"?  dso2 is not used in this test.

mwoodiupui reviewed (commented):
Other than one suspected typo, I think this is ready.

hardyoyo commented:
I've tested this PR in vagrant-dspace and can confirm that collection and community names properly sort in an alphabetical order. @mwoodiupu has some good points about a few things, I'd not have any problem with approving this code after Mark's concerns have been addressed.

hardyoyo commented:
Based on my testing, I'm going to approve this, however, I will wait for @mwoodiupui to merge it, as he seems to have concerns, still.

pnbecker commented on dspace-api/src/main/java/org/dspace/content/comparator/NameAscendingComparator.java:
```diff
@@ -0,0 +1,29 @@
+        }else if (dso2 == null){
+            return 1;
+        }else {
+            return ObjectUtils.compare(dso1.getName(),dso2.getName());
```
I would suggest to either use `StringUtils.compareIgnoreCases(String, String)` here or to trim the Strings and convert them to either small or capital letters. Currently sorting the community/collections ends up in listing all communities/collection A-Z starting with capital letters first and then listing the communities/collection a-z starting with small letters. I would expect one alphabetical sequence, not two.

pnbecker reviewed (changes requested):
I tested this in JSPUI and it works fine. There is just one issue left: currently it sorts communities/collections with a strange respect for small/capital letters.

Commit:
DS-3406: Resolve review feedback

tomdesair commented:
Hi @pnbecker and @mwoodiupui 
My last commit addresses your comments: https://github.com/DSpace/DSpace/pull/1684/commits/e358cb84d1c4ade5056f3c35912adfddab660eec

Thanks for the thorough review!

mwoodiupui reviewed (approved):
+1 by inspection.  My concerns are addressed, and overall it looks well.

hardyoyo reviewed (approved):
+1 by review, and the tests were also revised to prove that compareToIgnoreCase does what it's supposed to do.

hardyoyo reviewed (approved):
Pascal's concerns have been addressed.

pnbecker reviewed (approved):
I haven't retested it, but the code looks good to me. Thanks a lot for the change!
+1

tdonohue commented:
This was cherry-picked to `master` in #1768
